### PR TITLE
dynamic cooldowns & other neato stuff

### PR DESCRIPTION
**Pull request by nsupc**
- users.get now returns a user or raises an error; no need to handle in commands
- queue.get_nations does the same
- checks/errors now take discord.User instead of an id or a users.User
- added custom errors for empty queue and recent recruitment (which includes a retry after value)
- the time that the next recruitment is allowed is now dynamic based on the number of nations recruited last

### CODE REVIEW

**Comment by brrscc** [comment]
Nice work, UPC! Looks good to me.

File: cogs/recruit.py [hunk 1]
```diff
@@ -139,12 +164,13 @@ async def newnations_polling(self):
             self.bot.std.info("Polling NEWNATIONS shard.")
 
             new_nations = bs(requests.get("https://www.nationstates.net/cgi-bin/api.cgi?q=newnations",
-                             headers=headers).text, "xml").NEWNATIONS.text.split(",")  # type: ignore -- BeautifulSoup returns a variant type.
```
Comment (Darcness) [inline cogs/recruit.py:142]: Can we put the explanatory comment back in please?

File: cogs/recruit.py [hunk 2]
```diff
@@ -139,12 +164,13 @@ async def newnations_polling(self):
...
+                                          headers=headers).text, "xml").NEWNATIONS.text.split(",")  # type: ignore
         except:
             # certified error handling moment
-            self.bot.std.error("An unspecified error occured while trying to reach the NS API")
+            self.bot.std.error("An unspecified error occurred while trying to reach the NS API")
```
Comment (Darcness) [inline cogs/recruit.py:169]: :)

File: cogs/recruit.py [hunk 3]
```diff
@@ -179,7 +205,8 @@ async def reporter(self):
         date = datetime.now().strftime("%Y-%m-%d")
 
         if reports_channel is not None:
-            await reports_channel.send(f"Daily Report: {date}\n```{summary}```")  # type: ignore -- Not sure why Pylance can't see this...
+            await reports_channel.send(
+                f"Daily Report: {date}\n```{summary}```")  # type: ignore
```
Comment (Darcness) [inline cogs/recruit.py]: revert, plz

File: cogs/sessions.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    async def session(self, ctx: commands.Context, interval: int = 45):
+        if self.sessions.get(ctx.author.id):
+            raise SessionAlreadyStarted(ctx.author)
+
+        self.bot.rusers.get(ctx.author).active_session = True
+
+        if interval < 45:
+            interval = 45
+
+        session = Session(self.bot, ctx.author, ctx.channel.id, interval)
```
Comment (Darcness) [inline cogs/sessions.py]: You're not using the 'session' variable here, may as well cram these two onto the same line?

File: components/users.py [hunk 5]
```diff
@@ -37,11 +46,11 @@ def add(self, user: User):
 
         self.save()
 
-    def get(self, id: int):
-        for user in self.users:
-            if user.id == id:
-                return user
-        return None
+    def get(self, user: discord.User) -> User:
```
Comment (Darcness) [inline components/users.py:49]: Not sure why we're bothering with this.  Fetching by ID was working fine, and we're not getting any value in sending down the entire object.

**Review by Darcness — APPROVED** [review]

File: components/users.py (hunk 5 above)
Comment (brrscc) [reply components/users.py:49]: It is required to use the Discord.py CommandError interface

**Pull request merged**